UsernamePasswordSecondFactor/UsernamePasswordSecondFactor/UsernamePasswordAdapter.cs

```diff
@@ -79,34 +80,43 @@ public IAdapterPresentation TryEndAuthentication(IAuthenticationContext authCont
                 throw new ExternalAuthenticationException(ResourceHandler.GetResource(Constants.ResourceNames.ErrorNoAnswerProvided, authContext.Lcid), authContext);
             }
 
-            string username = (string)proofData.Properties[Constants.PropertyNames.Username];
-            string password = (string)proofData.Properties[Constants.PropertyNames.Password];
-
-            // validate that the username posted back matches the identity we saved in the encrypted blob.
-            // this should never fail for valid requests.
+            if (!authContext.Data.ContainsKey(Constants.AuthContextKeys.Identity))
+            {
+                Trace.TraceError(string.Format("TryEndAuthentication Context does not contains userID."));
+                throw new ArgumentOutOfRangeException(Constants.AuthContextKeys.Identity);
+            }
+            
             if (!authContext.Data.ContainsKey(Constants.AuthContextKeys.Identity))
             {
                 throw new ArgumentNullException(Constants.AuthContextKeys.Identity);
             }
+            
+            string username = (string)authContext.Data[Constants.AuthContextKeys.Identity];

```

Comment:
Is it okay to take this directly from the authContext.Data without checking it against the proofData.Properties like you were doing before?

Reply:
yes, authcontext is encrypted so it's safe to trust.